subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -92,20 +94,43 @@ private SnapshotHierarchy updateRootNotifyingWatchers(SnapshotHierarchy currentR
     }
 
     @Override
-    public void afterBuildStarted(boolean watchingEnabled) {
+    public void afterBuildStarted(boolean watchingEnabled, BuildOperationRunner buildOperationRunner) {
         reasonForNotWatchingFiles = null;
-        rootReference.update(currentRoot -> {
-            if (watchingEnabled) {
-                SnapshotHierarchy newRoot = handleWatcherRegistryEvents(currentRoot, "since last build");
-                if (watchRegistry == null) {
-                    newRoot = startWatching(newRoot);
+        rootReference.update(currentRoot -> buildOperationRunner.call(new CallableBuildOperation<SnapshotHierarchy>() {
+            @Override
+            public SnapshotHierarchy call(BuildOperationContext context) {
+                if (watchingEnabled) {
+                    SnapshotHierarchy newRoot = currentRoot;
+                    boolean startedWatching = false;
+                    FileWatcherRegistry.FileWatchingStatistics statistics = watchRegistry == null
+                        ? null
+                        : watchRegistry.getAndResetStatistics();
+
+                    newRoot = stopWatchingIfProblemsWhenReceivingFileChanges(newRoot, statistics);
+                    if (watchRegistry == null) {
+                        startedWatching = true;
+                        context.setStatus("Starting file system watching");
+                        newRoot = startWatching(newRoot);
+                    }
+                    context.setResult(new BuildStartedFileSystemWatchingResult(
+                        true,
+                        startedWatching,
+                        statistics == null ? null : new DefaultFileSystemWatchingStatistics(statistics, newRoot))
+                    );
```

Comment:
I think inlining this type as an anonymous inner type would be much more self-documenting. Same for the other result type.

Reply:
Inlned.